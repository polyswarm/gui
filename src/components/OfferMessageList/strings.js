const strings = {
  empty: 'No messages have been sent. Send some files to get started.',
  payment: 'Payment',
  request: 'Request',
  assertion: 'Assertion',
  malicious: 'Malicious',
  safe: 'Safe',
};
export default strings;