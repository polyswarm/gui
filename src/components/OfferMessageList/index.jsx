import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
// Bounty imports
import Card from '../Card';
import CardContent from '../CardContent';
import CardHeader from '../CardHeader';
import StatRow from '../StatRow';
// Component imports
import strings from './strings';

class OfferMessageList extends Component {

  render() {
    const { props: { offer } } = this;
    const messages = offer.messages || [];
    return (
      <div className='OfferMessageList'>
        <ul>
          {messages.map((message) => {
            let card = null;
            if (message.type === 'request') {
              card = this.renderRequest(message);
            } else if (message.type === 'assertion') {
              card = this.renderAssertion(message);
            } else {
              // payment
              card = this.renderPayment(message);
            }
            return (card);
          })}
        </ul>
        {( messages.length === 0) && (
          <div className='OfferMessageList-Placeholder'>
            <h3>
              {strings.empty}
            </h3>
          </div>
        )}
      </div>
    );
  }

  renderRequest(message) {
    const files = message.files;
    const filesList = files.map((file) => file.name).reduce((accumulator, name) => `, ${name}`);
    return(
      <Card key={message.guid}>
        <CardHeader
          title={strings.request}
        />
        <CardContent>
          <ul>
            <StatRow title={strings.files}
              content={filesList}/>
          </ul>
        </CardContent>
      </Card>
    );
  }

  renderAssertion(message) {
    const files = message.files || [];
    const verdicts = message.verdicts || [];
    const worstVerdict = verdicts.reduce((accumulator, verdict) => accumulator || verdict);
    const verdictClass = classNames({
      'Assertion-Malignant': worstVerdict,
      'Assertion-Benign': !worstVerdict
    });
    return(
      <Card key={message.guid}>
        <CardHeader
          className={verdictClass}
          title={strings.assertion}
        />
        <CardContent>
          <ul>
            <StatRow title={strings.metadata}
              content={message.metadata}/>
            {files && files.map((file, index) => {
              return (
                <StatRow 
                  key={file.name}
                  title={file.name}
                  content={verdicts[index] ? strings.malicious : strings.safe}/>
              );
            })}
          </ul>
        </CardContent>
      </Card>
    );
  }

  renderPayment(message) {
    return(
      <Card key={message.guid}>
        <CardHeader
          title={strings.payment}
          subhead={`${message.amount}${strings.nectar}`}
        />
      </Card>
    );
  }
}

OfferMessageList.proptypes = {
  offer: PropTypes.object.isRequired,
};
export default OfferMessageList;
