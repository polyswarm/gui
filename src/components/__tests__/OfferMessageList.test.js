import React from 'react';
import {render} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import OfferMessageList from '../OfferMessageList';

const offer = {
  author: 'author',
  expert: 'expert',
  remaining: '100',
  closed: false,
  messages: [
    {
      guid: 'asdf',
      type: 'payment',
      amount: '.5'
    },
    {
      guid: 'fdsa',
      type: 'payment',
      amount: '.0625'
    },
  ]
};

const walletList = [
  {address:'author', nct: '1', eth: '1'},
  {address:'demo', nct: '1', eth: '1'},
  {address:'omed', nct: '1', eth: '1'}
];

it('renders without crashing', () => {
  const wrapper = render(<OfferMessageList 
    offer={offer}
    walletList={walletList}/>);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});