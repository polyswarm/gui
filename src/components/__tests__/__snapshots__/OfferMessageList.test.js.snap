// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="OfferMessageList"
>
  <ul>
    <li>
      <div
        class="Card"
      >
        <header
          class="CardHeader"
        >
          <div
            class="CardHeader-Title"
          >
            Payment
            <p
              class="CardHeader-Sub"
            >
              .5 Nectar (NCT)
            </p>
          </div>
        </header>
      </div>
    </li>
    <li>
      <div
        class="Card"
      >
        <header
          class="CardHeader"
        >
          <div
            class="CardHeader-Title"
          >
            Payment
            <p
              class="CardHeader-Sub"
            >
              .0625 Nectar (NCT)
            </p>
          </div>
        </header>
      </div>
    </li>
  </ul>
</div>
`;
